add optional field support for degenObject

LoganBarnett opened the pull request:
There's a few breaking changes in here. `degenObject` now requires, ironically, a list of optional fields. `codeGen` has added a parameter for easier debugging that disables `prettier` so you can still debug the generated output even with a syntax error. Any functions depending on `codeGen` will also need to supply or chain along the value.

LoganBarnett committed:
add optional field support for degenObject

There's a few breaking changes in here. degenObject now requires, ironically, a
list of optional fields. codeGen has added a parameter for easier debugging that
disables prettier so you can still debug the generated output even with a syntax
error. Any functions depending on codeGen will also need to supply or chain
along the property.

gyrfalcon commented on src/generator.js:
```diff
@@ -118,6 +141,22 @@ export const degenObject = <CustomType: string, CustomImport: string>(
+}
+
+
+export const degenOptionalField = <CustomType: string, CustomImport: string>(
```
It may be worth calling out in the documentation when one would use `degenOptionalField` over `degenField`, especially since the example doesn't use `degenOptionalField` at all.

LoganBarnett replied:
Ugh - it's vestigial. I'll make sure it's removed.

LoganBarnett replied:
Removed.

gyrfalcon commented on test/merge-deps.js:
```diff
@@ -0,0 +1,34 @@
+  type DeserializerGenerator,
+  type MetaType,
+  typeHeader,
+} from '../src/generator.js'
```
I've been noodling on the "forgot to add an export to index.js" problem, and I think having the tests import from `index.js` may be the simplest thing to do for now.

Also, as a related issue, new tests need to be added to `test-suite.js`, which I can now see will have a similar problem of being forgotten. Mayhaps `test-suite.js` can be smartened up to slurp in all the files in the test directory, or a real test framework needs to be brought in.

LoganBarnett replied:
Re: tests - I'm thinking there's enough itch that a real framework would scratch it.

Re: `index.js` I think mandating that tests use that is very sensible.

gyrfalcon replied:
Something else I just noticed, `typeHeader` isn't currently exported by `../src/generator.js`.

LoganBarnett replied:
Good catch! Also I was hoping to do a big pass on `typeHeader` and making sure that type names with their type parameters are respected throughout the generation functions. But it'll probably be another pull request.

LoganBarnett replied:
I've exported `typeHeader` but still want to take a comprehensive pass at documenting it and using it everywhere in a later pull request.

gyrfalcon commented:
I've temporarily removed this comment because I realized I'm a dumb dumb and there may not actually be a problem with only-optional objects.

gyrfalcon commented:
> I've temporarily removed this comment because I realized I'm a dumb dumb and there may not actually be a problem with only-optional objects.

Okay, I have confirmed there's no bug with only-optional objects. The call to `fileGen` in `flow-degen.js`, however,  needs to be updated to pass in the new `prettier: bool` parameter.

gyrfalcon commented:
> > I've temporarily removed this comment because I realized I'm a dumb dumb and there may not actually be a problem with only-optional objects.
> 
> Okay, I have confirmed there's no bug with only-optional objects. The call to `fileGen` in `flow-degen.js`, however, needs to be updated to pass in the new `prettier: bool` parameter.

Never mind, I'm still a dumb dumb (not a great morning for smart from me it seems), the optional-only object issue still exists. I had put in a temporary required field to verify that that fixed the issue, but then forgot to remove the temporary field before retesting. Here's the original comment again:

In doing some work with this branch, it looks like there's a bug if you have an object with only optional fields, e.g.:

```javascript
type NothingDefinite = {
  foo?: string,
}

...

const generateRefineNothingDefinite = () => degenObject(
  { name: 'NothingDefinite', typeParams: [] }, [], 
    [
      degenField('foo', degenString()),
    ]
)
```

Which results in an error when prettier is tidying up the generated code:

```
SyntaxError: Unexpected token, expected "," (125:1)
  123 |     }
  124 |   }
> 125 | }
      | ^
  126 | ),
```

LoganBarnett committed:
fix degenObject required and optional fields

When mixing optional fields with required fields, a brace mismatch would occur
that has now been fixed. This adds a test to capture the optional-only case as
well as the required + optional case. We now assign all fields to the "unsealed"
object because if we add required fields upon object creation they will be
unsealed and thus optional fields will fail a type check.

Some quality of life issues with testing have been addressed.

LoganBarnett committed:
remove vestigial degenOptionalField and export typeHeader

LoganBarnett reviewed (commented):
Issues regarding mixing/matching optional + required should be addressed now.

LoganBarnett committed:
fix added test from the rebase and announce typeHeader

gyrfalcon commented:
My comment about `flow-degen.js` needing to be updated to pass the new `prettier: bool` paramenter to `fileDegen` still applies. It probably got lost it the shuffle of my back-and-forthing about the mixed optional/non-optional issue.

LoganBarnett committed:
fix lack of type coverage in flow-degen.js and really export typeHeader

fileGen's reference comes from dist, which has no real type information. That
said we can import the "typeof" fileGen from its original source file and cast
the type there. This requires us to fix the lack of prettify argument, and
should prevent similar misses going forward.

LoganBarnett commented:
@gyrfalcon the missing `prettier` arg wasn't caught in the type check, so it slipped by. Good catch! I made a modification to catch future issues that are just like it in that file.

gyrfalcon commented:
Looks good!
